Introduce new approach for authentication utilising ```AzureCliCredential```

valerii-martell opened the pull request:
Closes https://github.com/equinor/fiberoptics-common/issues/46

Started in https://github.com/equinor/ngrmdf-fotone-application-framework/pull/386

Pre-requisite for:
- https://github.com/equinor/fiberoptics-fotonepy/pull/207
- https://github.com/equinor/ngrmdf-fotone-integration-tests/pull/80
- https://github.com/equinor/fiberoptics-processing-prototype/issues/8
- And eventually for https://github.com/equinor/fotone-application-library

valerii-martell commented on fiberoptics/common/auth/__init__.py:
```diff
@@ -5,14 +5,22 @@
     InteractiveBrowserCredential,
     TokenCachePersistenceOptions,
+    DefaultAzureCredential,
 )
```
[Use async credential](https://github.com/equinor/fiberoptics-common/pull/47/commits/166913b07ec0df86170a6c8fe35315d9580110fa)

valerii-martell commented on fiberoptics/common/auth/__init__.py:
```diff
@@ -213,3 +221,55 @@ def remove_cached_credential(name: str):
+                "exclude_visual_studio_code_credential": True,
+                "exclude_interactive_browser_credential": True,
+            }
+            cls._credential = DefaultAzureCredential(**options)
```
[Use async credential](https://github.com/equinor/fiberoptics-common/pull/47/commits/166913b07ec0df86170a6c8fe35315d9580110fa)

trylle commented:
1. Something highly unfortunate that I discovered with `AzureCliCredential`, is that it does _not_ cache tokens and IIRC invokes a potentially expensive az cli process for every token generated. See https://learn.microsoft.com/en-us/python/api/azure-identity/azure.identity.azureclicredential?view=azure-python#azure-identity-azureclicredential-get-token, https://github.com/Azure/azure-sdk-for-python/issues/40636#issuecomment-2819608804.
This makes it unsuitable for code that invokes many authorized requests - at least without adding caching ourselves. Do you have any citation for the claim that it does memoize tokens as described in #46?
2. Does using async credentials from a sync context work without any issues/changes to user code?
3. Is there any particular reason to exclude environment credentials? I don't see how this introduces any detriment, and it could be useful for testing service principals.

trylle commented on fiberoptics/common/auth/__init__.py:
```diff
@@ -213,3 +222,163 @@ def remove_cached_credential(name: str):
+                "exclude_interactive_browser_credential": True,
+            }
+            cls._credential = DefaultAzureCredential(**options)
+            if not cls._credential:
+                raise NoCredentialsAvailable("No credentials are available")
```
How can this happen?

valerii-martell replied:
Resolved in [Dismiss unnecessary null check](https://github.com/equinor/fiberoptics-common/pull/47/commits/9f9b4e364e4bbbeaed2e75307c9a9995e860dd0c)

valerii-martell commented on fiberoptics/common/auth/__init__.py:
```diff
@@ -213,3 +222,163 @@ def remove_cached_credential(name: str):
+        Credential
+            The singleton credential instance for the specified resource_id.
+        """
+        if resource_id not in cls._instances:
+            instance = super().__new__(cls)
+            cls._instances[resource_id] = instance
+        return cls._instances[resource_id]
```
Yes, I agree, ideally it will be like you're saying. 

However, the usecase I had in mind - when we use two libraries alongside each other each using fiberoptics-common for authentication. Like the framework (after we adopt these changes there) and fotonepy, or the library and fotonepy, integration tests and fotonepy, and so on. As long as the framework or the library use its own internal Fotone client implementation, not fotonpy, we'll have two credential objects in the runtime pointing basically to the same scope (Fotone API), at least I assume that's what will happen. So caching on the fiberoptics-common should help to prevent that. When we update internal clients so their requests go through fotonepy then of course no need for this caching.

Although it is a very narrow use case indeed and I was also curious to hear your feedback on this actually :)

valerii-martell commented:
> Does using async credentials from a sync context work without any issues/changes to user code?

It does, as far as I can see, on the fotonepy side I tested it with both sync and async clients. The main use case I had in mind was fotonepy, there I took an inspiration from the way how async and sync clients are implemented. As you can see, only the async clients there have real implementation, while the sync ones just run synchronously their async counterparts https://github.com/equinor/fiberoptics-fotonepy/blob/c39cbdcd7a5c5bb71ffc155b4ce80c5096fde604/fiberoptics/fotonepy/clients/app.py#L452. 

In terms of code it integrates seamlessly with async clients, where you just pass Credential objects, not the token itself. The only added tiny complexity was here in our custom implementation, where we directly attach token to the request to api, and therefore we need to await the token now https://github.com/equinor/fiberoptics-fotonepy/blob/9af1b69f643d053de7041049a64b5a6a173ba7fa/fiberoptics/fotonepy/core.py#L64

Fotonepy had the only one sync client - BlobClient, but I rewrote it to Async and added a sync counterparts following the logic of all the other clients there. All of this is in that fotonepy PR assigned to you https://github.com/equinor/fiberoptics-fotonepy/pull/207
